Work on table padding issue #95

IDisposable opened the pull request:
## TARGETS `master` for the  **v2-release-chain**

Changed `getDefaultStyle` to create the element in the sandbox so it is pure (not tainted by CSS / ShadowDOM in the main document).
Never clone SCRIPT elements (that's just asking for trouble).
Removed `imagediff` package as it's not useful anymore.

IDisposable committed:
Work on table padding issue #95

Changed `getDefaultStyle` to create the element in the sandbox so it is pure (not tainted by CSS / ShadowDOM in the main document).
Never clone SCRIPT elements (that's just asking for trouble).
Removed `imagediff` package as it's not useful anymore.

IDisposable commented:
@meche-gh can you see if this is makes sense to you?

I THINK the problem is that we're creating the element for the `tagName` to extract the default styles in our actual `window.contentWindow.document`, not in the *sandbox*.

This means that global styles that are applied to the tags like a css rule `td { padding: 0px; }` in the actual DOM will copy those "not really default" properties over to the sandboxed element. Later, when we compare the source element's style it looks unchanged from the defaults, but that's not true.

Rather, I think should be creating the element against the `sandbox.contentWindow`'s document so the properties are not tainted and reflect what the browser would really have defaulted to.

meche-gh commented:
Sounds good to me. I noticed that the iframe getComputedStyle is a little slower (i.e. forcing a layout recompute within the iframe) - if that's a needed investment then this is the way

IDisposable commented:
Much playing around later, it's not possible to fix because something is keeping the `padding: 0px;` from making it to the SVG, even though the values are correctly set on the cloned node. I've verified this by adding to the original fiddle leveraging the ability to insert the cloned node into the existing page (and it looks IDENTICAL because all the same styles are applying)
See https://jsfiddle.net/Lrgkwt5m/87/

IDisposable committed:
Add ability to insert the cloned node into the test

The allows inspecting the properties of the clone.

IDisposable committed:
Simplified the SVG

Removed the setting of the `top` and `left`
Set the `width` and `height` on the `foreignObject` element to the target size.
Removed setting of `width` and `height` on the `svg` element.

IDisposable committed:
Copy the size from the original object when possible

Falls back to the old `scrollWidth`/`scrollHeight` logic if the `width` and `height` aren't set in `px`.

IDisposable commented:
Pushed the final state of the v2 based code cleanup. Still doesn't fix the original bug (because I think it's a browser issue), but this code is cleaner and less likely to break.

IDisposable commented:
@meche-gh feel free to make any comments before I merge this tomorrow.

meche-gh commented on src/dom-to-image-more.js:
```diff
@@ -430,10 +436,10 @@
             })
             .then(function (foreignObject) {
-                return `<svg xmlns="http://www.w3.org/2000/svg" width="${width}" height="${height}">${foreignObject}</svg>`;
+                return `<svg xmlns="http://www.w3.org/2000/svg">${foreignObject}</svg>`;
```
I personally think we should set a viewport here:

```suggestion
                return `<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 ${width} ${height}">${foreignObject}</svg>`;
```

meche-gh replied:
Or we could do both! https://github.com/whatwg/html/issues/3510#issuecomment-369982529 - seems like SVGs just need both for good cross-browser compat..

IDisposable replied:
`viewBox` is problematic [because those values are floats, not integers.](https://github.com/whatwg/html/issues/3510#issuecomment-370101075)

meche-gh replied:
This lib would be generating SVGs without a bounding box. I consider this to be "canvas mode" - it's an image but without any notion of natural dimensions. Less useful and likely a major change in semver.
It's more convenient for devs to pass the natural dimension matching the original element. `width="${width}" height="${height}"` is valuable imho

IDisposable replied:
I've restore the width and height on the SVG image.

IDisposable commented:
Cool, I'll revert that part.

On Fri, Jan 13, 2023 at 2:49 AM meche-gh ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/dom-to-image-more.js
> <https://github.com/1904labs/dom-to-image-more/pull/104#discussion_r1069098099>
> :
>
> >              })
>              .then(function (foreignObject) {
> -                return `<svg xmlns="http://www.w3.org/2000/svg" width="${width}" height="${height}">${foreignObject}</svg>`;
> +                return `<svg xmlns="http://www.w3.org/2000/svg">${foreignObject}</svg>`;
>
> This lib would be generating SVGs without a bounding box. I consider this
> to be "canvas mode" - it's an image but without any notion of natural
> dimensions. Less useful and likely a major change in semver.
> It's more convenient for devs to pass the natural dimension matching the
> original element. width="${width}" height="${height}" is valuable imho
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/1904labs/dom-to-image-more/pull/104#discussion_r1069098099>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAGVMSYR2OKDAWE7DONEDLLWSEJJVANCNFSM6AAAAAATY4HMTY>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 
Marc Brooks
http://about.me/IDisposable
http://twitter.com/IDisposable

IDisposable committed:
Revert to have dimensions on SVG object.

IDisposable commented:
Had a thought last night at the hockey game (the Blues were being terrible in the third).

What if the issue is that at the time the styles are being copied over, the destination TH/TD has all the default properties so we don't mechanistically set the padding. Then later in the process we set the `table-cell` style and that's when the 1px padding is kicking in... if this is the case, I would have to figure out the right order to copy style properties over in... that sounds like a nightmare.

meche-gh commented:
[THEY'Re fixing it!!](https://bugs.chromium.org/p/chromium/issues/detail?id=1406929#c6) :tada:

meche-gh commented:
The only way to *bypass* the Chromium issue, and also support the default styling on descendants of `<ruby>`, is to make `defaultElement` wrapped recursively with the parents of `targetElement` in `getDefaultStyle`.

We would iterate upwards on the DOM tree and get all the parent tag names going up to `document.documentElement` - that would definitely futureproof it. This would make caching much more complex however.

IDisposable commented:
This sounds like an excellent path forward... I'm going to merge where we are to get back to sanity on the other PRs. Issue #106 capture the remaining issue and expected solution.
